src/processing/BoardState.java

```diff
@@ -0,0 +1,211 @@
+		posToObjects = new HashMap <Position, PriorityQueue <PriorityQueueObject>> ();
+	}
+
+	/**
+	 * Synchronized handles to get and set the 'selectedObject' data member in ClientBoardState class
+	 */
+	public synchronized void setSelectedObject(PriorityQueueObject pqObject) {
+		ClientBoardState.selectedObject = new PriorityQueueObject(pqObject);
+	}
+
+	public synchronized PriorityQueueObject getSelectedObject() {
+		return ClientBoardState.selectedObject;
+	}
+
+	/**
+	 * Looks up BoardObject for the ObjectId key in the argument
+	 */
+	public synchronized BoardObject getBoardObjectFromId(ObjectId objId) {
+		return objIdToBoardObject.get(objId);
+	}
+
+	/**
+	 * Gets ObjectIds at the top of Priority Queues at each position (x, y) in given ArrayList.
+	 * The ObjectId that is present at the maximum number of positions i.e. which occurs
+	 * the maximum number of times is returned
+	 * This is done as the selected pixel positions can be such that
+	 * some of the pixels belong to an object and some to others.
+	 */
+	public synchronized ObjectId getMostProbableObjectId(ArrayList <Position> positions) {
+		HashMap<ObjectId, Integer> count = new HashMap<ObjectId, Integer>();
+		for (Position pos: positions) {
+			PriorityQueue <PriorityQueueObject> queue_at_pos = posToObjects.get(pos);
+			if (queue_at_pos == null) {
+				continue;
+			}
+			PriorityQueueObject top_obj = (PriorityQueueObject) queue_at_pos.peek();
+			if (top_obj == null) {
+				continue;
+			}
+
+			Integer curr = count.get(top_obj.objectId);
+			if (curr == null) {
+				curr = 0;
+			}
+			count.put(top_obj.objectId, curr + 1);
+		}
+
+		Integer maxOccurrence = 0;
+		ObjectId maxOccObjectId = null;
+		Iterator <Map.Entry<ObjectId, Integer>> it = count.entrySet().iterator();
+
+		while (it.hasNext()) {
+			Map.Entry <ObjectId, Integer> mapElement = (Map.Entry <ObjectId, Integer>) it.next();
+			if (maxOccurrence < mapElement.getValue()) {
+				maxOccurrence = mapElement.getValue();
+				maxOccObjectId = mapElement.getKey();
+			}
+		}
+
+		return maxOccObjectId;
+	}
+
+	/**
+	 * For each Position (x, y) in the input ArrayList, insert the PriorityQueueObject into the
+	 * Priority Queue at the position (x, y) present in the map
+	 */
+	private synchronized void insertIntoPQ (ArrayList <Position> positions, PriorityQueueObject toBeInserted) {
+		for (Position pos : positions) {
+			PriorityQueue <PriorityQueueObject> pq = posToObjects.get(pos);
+			if (pq == null) {
+				posToObjects.put(pos, new PriorityQueue<PriorityQueueObject>(20, new ObjectComparator()));
+			}
+			posToObjects.get(pos).add(toBeInserted);
+		}
+	}
+
+	/**
+	 * For each Position (x, y) in the input ArrayList, delete the PriorityQueueObject from the
+	 * Priority Queue at the position (x, y) present in the map, if it exists in the Priority Queue
+	 */
+	private synchronized void removeFromPQ (ArrayList <Position> positions, PriorityQueueObject objectToBeRemoved) {
```

Comment:
Can you optimize it a bit, like if a priority queue becomes empty, you remove it from the `posToObjects` hash map ?

Reply:
@wildcat26